grub-core/fs/zfs/zfs.c

```diff
@@ -3809,13 +3810,49 @@ grub_zfs_envblk_open (struct grub_file *file)
 			       sizeof (vdev_boot_envblock_t));
       if (err == GRUB_ERR_NONE)
 	{
+	  size_t size = sizeof (vbe->vbe_bootenv);
+	  char *buf = grub_malloc (size);
 	  vbe = (vdev_boot_envblock_t *)data->file_buf;
-	  if (vbe->vbe_version != VB_RAW)
+	  if (vbe->vbe_version == VB_RAW)
 	    {
-	      zfs_unmount (data);
-	      return GRUB_ERR_BAD_FILE_TYPE;
+	      grub_strncpy (buf, vbe->vbe_bootenv, size - 1);
 	    }
-	  file->size = grub_strnlen (vbe->vbe_bootenv, sizeof (vbe->vbe_bootenv));
+	  else
+	    {
+	      size_t count;
+	      char *nvl = vbe->vbe_bootenv;
+	      const char *nvp = NULL;
+	      grub_snprintf (buf, size, "%s", GRUB_ENVBLK_SIGNATURE);
+	      grub_memset (buf, '#', size);
```

Comment:
`grub_snprintf` should be called after `grub_memset`, otherwise we are overwriting `GRUB_ENVBLK_SIGNATURE`